docs/typescript/more.md

```diff
@@ -0,0 +1,195 @@
+
+## Avoid `any`
+
+The `any` type is dangerous and should be avoided. It makes it impossible for the type system to
+discover issues or provide helpful information in many cases.
+
+`unknown` is a safer alternative to `any` that may be helpful.
+
+```ts
+// ⛔️ any is dangerous. Do not use it!
+const x: any = 0;
+const y: string = x; // The type system thinks this is fine! 😱
+
+// ✅ unknown is a good alternative to any.
+const x2: unknown = produceSomeUntypedExternalValue();
+const y2: string = x2; // 😌 Type error. The type system correctly doesn't trust this to be a string.
+```
+
+## Avoid type assertions
+
+A _type assertion_ looks like this: `const x = 42 as string` or `const x = <string>42`, although the
+latter form is less common due to the similar or ambiguous syntax with JSX. Some folks call these
+"casts" or "coercion."
+
+Type assertions are very dangerous and should be avoided. They override the type system and tell it
+that we know better.
+
+If you need to add a type to something, instead of an assertion try an _annotation_, like this:
+`const x: number = 42`. This is help for TypeScript, it will use the annotated type but will also
+check that it is valid. But don't annotate an assignment like this, TypeScript will infer it 🙂
+
+## Use readonly
+
+We often rely on immutable data, we can encode this in the type system in a few ways.
+
+First, when defining an interface we can be clear about readonly properties to make our interface
+immutable:
+
+```ts
+interface CantTouchThis {
+	readonly string: string;
+	readonly array: ReadonlyArray< string >;
+	readonly tuple: readonly [ string, string ];
+	readonly object: {
+		readonly prop: number;
+	};
+}
+```
+
+This also works in classes:
+
+```ts
+class Klassy {
+	readonly immutable: string;
+	mutable: number;
+
+	constructor() {
+		this.immutable = 'define it here, last chance';
+		this.mutable = 0;
+	}
+
+	tick() {
+		this.mutable += 1; // ✅ Ok
+		this.immutable = 'whoops'; // ⛔️ Type error, that's readonly!
+	}
+
+	// Yes, these arrow functions are instance properties too and can be readonly 👀
+	readonly noop = () => undefined;
+}
+```
+
+And did you see that `ReadonlyArray<T>`? That's a great alternative to `Array<T>` or `T[]`. There
+are also `ReadonlySet<T>` and `ReadonlyMap<Key, Value>` types!
+
+## Use const assertions
+
+Not everything needs a _const assertion,_ but when we want to infer a readonly interface it's a
+great option. Here's what that might look like:
+
+```ts
+function actionMaker( direction: 'up' | 'down' ) {
+	return { type: 'GO', direction } as const; // Here is our const assertion: `as const`
+}
+```
+
+This function has the following return type, inferred without writing a bunch of types. Perfect!
+We've leveraged inference nicely:
+
+```ts
+type ReturnType = {
+	readonly type: 'GO';
+	readonly direction: 'up' | 'down';
+};
+```
+
+## Avoid enums
+
+Avoid enums, they don't add much over simpler alternatives.
```

Comment:
This originally linked to an internal posts. I can get that post published somewhere more visible so it can be linked.

Reply:
That will be quite nice 🙂 Although, we can probably add a short TL;DR for those who want a quick overview.

Reply:
Added in b3e9870